<div style="text-align: center;">
  <img src="Images/Weather_Types.png" alt="Weather_Types Illustration" width="600"/>
</div>

## Hello!  

This is a project utilizing the **Weather-Type-Classification Database** to analyze and classify weather patterns. By exploring this dataset, we aim to build insights into weather characteristics and develop models for accurate weather classification based on the provided features.  

#### About the Dataset  
The dataset, available on [Kaggle](https://www.kaggle.com/datasets/nikhil7280/weather-type-classification), contains synthetically generated weather data designed for classification tasks. It includes a wide range of weather-related variables, offering opportunities to practice data preprocessing, feature engineering, and outlier detection. 

#### Dataset Overview  
The dataset consists of **13,200 rows** and **11 columns**, structured as follows:  

1. **Temperature**: Temperature in degrees Celsius, ranging from extreme cold to extreme heat.  
2. **Humidity**: Humidity percentage, including values above 100% to introduce outliers.  
3. **Wind Speed**: Wind speed in kilometers per hour, including unrealistically high values.  
4. **Precipitation (%)**: Precipitation percentage, with some outlier values.  
5. **Cloud Cover**: Description of cloud cover (categorical).  
6. **Atmospheric Pressure**: Atmospheric pressure in hPa, covering a wide range.  
7. **UV Index**: Strength of ultraviolet radiation (numeric).  
8. **Season**: Season during which the data was recorded (categorical).  
9. **Visibility (km)**: Visibility in kilometers, with very low or very high values.  
10. **Location**: The type of location where the data was recorded (categorical).  
11. **Weather Type**: Target variable, classifying weather as Rainy, Sunny, Cloudy, or Snowy.  

#### Project Inspiration  
This project is inspired by the need to develop robust classification models for weather prediction. By leveraging the **Weather-Type-Classification Database**, we aim to identify key relationships among variables and improve the accuracy of weather categorization.  

#### Goals of the Project  
1. **Data Analysis**: Explore the dataset to uncover trends and patterns in weather features.  
2. **Classification Models**: Develop and evaluate machine learning models to classify weather types based on the provided features.  

This project will focus on utilizing the **Weather-Type-Classification Database** to provide valuable insights into weather data and enhance classification strategies for weather-related applications.

> ⚠️ *Note*: Although the data is synthetic, it serves as a useful proxy for learning data preprocessing, outlier detection, and classification modeling techniques.

# **Step 3: Exploratory Data Analysis (EDA) – SQL Queries**

This notebook focuses on performing **exploratory data analysis using SQL queries** within a Jupyter environment using SQLite. By leveraging SQL-style queries, we can extract meaningful insights from the weather dataset efficiently and in a structured manner.

---

### Objectives of This Notebook

1. [Import Libraries and Load the Dataset](#import)  
2. [Preview the Dataset](#preview)  
3. [Run SQL Queries to Answer Key Questions](#queries)  
4. [Summary and Key Insights](#summary)

---

### Key Questions Explored via SQL

1. What is the frequency distribution of weather types such as Rainy, Cloudy, and Sunny?  
2. How does the average temperature vary across different seasons?  
3. What are the minimum and maximum humidity levels under each weather type?  
4. Which cloud cover category appears most frequently in the dataset?  
5. How does UV Index differ across locations like coastal, inland, and mountain areas?  
6. What are the average precipitation patterns across different seasons and weather types?  
7. How does wind speed behave on rainy days compared to other weather conditions?  
8. Which season has the lowest average visibility, and how does it relate to weather type?  
9. What are the atmospheric pressure trends across different locations and cloud cover categories?

---

### Next Steps

- Step 4: [Modeling & Prediction](./04_modeling_prediction.ipynb)


<a id="import"></a>

## **3.1 Import Libraries, Load the Dataset, and Connect to the Database**

We begin by importing the required Python libraries:

- **Pandas** for data manipulation and analysis.
- **SQLite3** for setting up and querying a local SQL database.
- **IPython SQL extension** for running SQL queries directly within the notebook.

After loading the **Student Depression** dataset into a pandas DataFrame, we connect it to a local **SQLite** database. This setup enables us to perform SQL-based exploration conveniently within the Jupyter Notebook environment.

In [1]:
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd

# sqlite3 is a built-in Python library for creating and interacting with SQLite databases
import sqlite3

# prettytable is a module used to format tabular data in a readable way (optional)
import prettytable

In [2]:
# Load the ipython-sql extension to run SQL queries in Jupyter
%load_ext sql

# Sets default format for prettytable display
prettytable.DEFAULT = 'DEFAULT'

In [3]:
# Create SQLite connection and cursor
con = sqlite3.connect("weather_classification_insights.db")
cur = con.cursor()

# Connect to the SQLite database for ipython-sql
%sql sqlite:///weather_classification_insights.db

In [4]:
# Load your cleaned dataset
df = pd.read_csv("weather_classification_cleaned.csv")

# Save the DataFrame to a new table in the SQLite database
df.to_sql("Climate_Analysis", con, if_exists='replace', index=False)

11944

In [5]:
# Drop the cleaned query table if it already exists (to avoid duplication)
%sql DROP TABLE IF EXISTS Weather_Data;

 * sqlite:///weather_classification_insights.db
Done.


[]

In [6]:
# Create a new cleaned SQL table with all non-null rows (modify as needed)
%sql CREATE TABLE Weather_Data AS SELECT * FROM Climate_Analysis;

 * sqlite:///weather_classification_insights.db
Done.


[]

---

<a id="preview"></a>

## **3.2 Preview the Dataset**

To understand the structure and content of the dataset, we start by viewing the first few rows using an SQL query:

In [7]:
%%sql
SELECT * 
FROM Weather_Data 
LIMIT 10;

 * sqlite:///weather_classification_insights.db
Done.


[(14.0, 73, 9.5, 82.0, 'partly cloudy', 1010.82, 2, 'Winter', 3.5, 'inland', 'Rainy'),
 (39.0, 96, 8.5, 71.0, 'partly cloudy', 1011.43, 7, 'Spring', 10.0, 'inland', 'Cloudy'),
 (30.0, 64, 7.0, 16.0, 'clear', 1018.72, 5, 'Spring', 5.5, 'mountain', 'Sunny'),
 (38.0, 83, 1.5, 82.0, 'clear', 1026.25, 7, 'Spring', 1.0, 'coastal', 'Sunny'),
 (27.0, 74, 17.0, 66.0, 'overcast', 990.67, 1, 'Winter', 2.5, 'mountain', 'Rainy'),
 (32.0, 55, 3.5, 26.0, 'overcast', 1010.03, 2, 'Summer', 5.0, 'inland', 'Cloudy'),
 (-2.0, 97, 8.0, 86.0, 'overcast', 990.87, 1, 'Winter', 4.0, 'inland', 'Snowy'),
 (3.0, 85, 6.0, 96.0, 'partly cloudy', 984.46, 1, 'Winter', 3.5, 'inland', 'Snowy'),
 (3.0, 83, 6.0, 66.0, 'overcast', 999.44, 0, 'Winter', 1.0, 'mountain', 'Snowy'),
 (28.0, 74, 8.5, 107.0, 'clear', 1012.13, 8, 'Winter', 7.5, 'coastal', 'Sunny')]

---

<a id="sql_queries"></a>

## **3.3 Run SQL Queries to Answer Key Questions**

Here we run a series of SQL queries to explore and gain insights from the dataset. Each query focuses on a specific question relevant to understanding factors related to depression.

### 3.3.1 Frequency Distribution of Weather Types

**Description:**
This query counts how often each type of weather (e.g., Sunny, Rainy, Cloudy) appears in the dataset. It helps us understand the most and least common weather conditions recorded across different locations and seasons.

In [8]:
%%sql
SELECT weather_type, COUNT(*) AS count 
FROM Weather_Data 
GROUP BY weather_type 
ORDER BY count DESC;

 * sqlite:///weather_classification_insights.db
Done.


[('Snowy', 3079), ('Cloudy', 3003), ('Rainy', 2986), ('Sunny', 2876)]

### 3.3.2 Average Temperature by Season

**Description:**
By calculating the average temperature for each season, this query reveals seasonal trends in temperature variation. It's useful for understanding how weather shifts between Winter, Spring, Summer, and Fall.

In [9]:
%%sql
SELECT season, ROUND(AVG(temperature), 2) AS avg_temperature 
FROM Weather_Data 
GROUP BY season 
ORDER BY avg_temperature DESC;

 * sqlite:///weather_classification_insights.db
Done.


[('Spring', 25.35), ('Autumn', 25.29), ('Summer', 25.27), ('Winter', 9.09)]

### 3.3.3 Min and Max Humidity by Weather Type

**Description:**
This query shows the minimum and maximum humidity levels for each weather type. It helps assess how different weather conditions (e.g., Rainy, Sunny) relate to the range of humidity levels experienced.

In [10]:
%%sql
SELECT weather_type, 
       MIN(humidity) AS min_humidity, 
       MAX(humidity) AS max_humidity 
FROM Weather_Data 
GROUP BY weather_type;

 * sqlite:///weather_classification_insights.db
Done.


[('Cloudy', 20, 109),
 ('Rainy', 20, 109),
 ('Snowy', 20, 109),
 ('Sunny', 20, 109)]

### 3.3.4 Most Frequent Cloud Cover Type

**Description:**
This query determines which type of cloud cover is recorded most frequently in the dataset. It provides insights into typical sky conditions in the observed areas.

In [11]:
%%sql
SELECT cloud_cover, COUNT(*) AS count 
FROM Weather_Data 
GROUP BY cloud_cover 
ORDER BY count DESC 
LIMIT 1;

 * sqlite:///weather_classification_insights.db
Done.


[('overcast', 5700)]

### 3.3.5 Average UV Index by Location

**Description:**
This query calculates the average UV index in different locations (coastal, inland, mountain). It helps identify which regions experience higher UV exposure, which could be relevant for public health or climate analysis.

In [12]:
%%sql
SELECT location, ROUND(AVG(uv_index), 2) AS avg_uv_index 
FROM Weather_Data 
GROUP BY location 
ORDER BY avg_uv_index DESC;

 * sqlite:///weather_classification_insights.db
Done.


[('coastal', 4.37), ('inland', 3.47), ('mountain', 3.41)]

### 3.3.6 Average Precipitation by Season and Weather Type

**Description:**
This detailed query breaks down average precipitation across combinations of seasons and weather types. It helps determine which combinations are associated with heavier or lighter rainfall.

In [13]:
%%sql
SELECT season, weather_type, ROUND(AVG(precipitation), 2) AS avg_precipitation 
FROM Weather_Data 
GROUP BY season, weather_type 
ORDER BY avg_precipitation DESC;

 * sqlite:///weather_classification_insights.db
Done.


[('Winter', 'Snowy', 76.57),
 ('Winter', 'Rainy', 76.15),
 ('Autumn', 'Rainy', 75.96),
 ('Spring', 'Rainy', 75.81),
 ('Summer', 'Rainy', 75.12),
 ('Autumn', 'Snowy', 63.0),
 ('Summer', 'Snowy', 57.41),
 ('Spring', 'Snowy', 53.82),
 ('Summer', 'Cloudy', 38.34),
 ('Winter', 'Cloudy', 38.05),
 ('Spring', 'Cloudy', 37.67),
 ('Autumn', 'Cloudy', 36.97),
 ('Autumn', 'Sunny', 19.39),
 ('Summer', 'Sunny', 18.98),
 ('Spring', 'Sunny', 17.49),
 ('Winter', 'Sunny', 16.82)]

### 3.3.7 Average Wind Speed on Rainy Days vs Others

**Description:**
This query compares the average wind speed during rainy weather to other types of weather. It helps understand whether wind tends to be stronger or weaker when it's raining compared to clearer conditions.

In [14]:
%%sql
SELECT 
  CASE 
    WHEN weather_type = 'Rainy' THEN 'Rainy' 
    ELSE 'Other' 
  END AS condition, 
  ROUND(AVG(wind_speed), 2) AS avg_wind_speed 
FROM Weather_Data 
GROUP BY condition;

 * sqlite:///weather_classification_insights.db
Done.


[('Other', 8.52), ('Rainy', 13.82)]

### 3.3.8 Season with Lowest Average Visibility and Its Weather Type

**Description:**
This query identifies the combination of season and weather type that results in the lowest average visibility. It is useful for understanding which conditions are most likely to impair visibility and potentially affect transportation or outdoor activities.

In [15]:
%%sql
SELECT season, weather_type, ROUND(AVG(visibility), 2) AS avg_visibility 
FROM Weather_Data 
GROUP BY season, weather_type 
ORDER BY avg_visibility ASC 
LIMIT 1;

 * sqlite:///weather_classification_insights.db


Done.


[('Winter', 'Snowy', 2.97)]

### 3.3.9 Atmospheric Pressure by Location and Cloud Cover

**Description:**
This query explores how atmospheric pressure varies across different locations and cloud cover types. It can help highlight potential relationships between pressure systems and local weather conditions.

In [16]:
%%sql
SELECT location, cloud_cover, ROUND(AVG(atm_pressure), 2) AS avg_pressure 
FROM Weather_Data 
GROUP BY location, cloud_cover 
ORDER BY avg_pressure DESC;

 * sqlite:///weather_classification_insights.db
Done.


[('mountain', 'clear', 1020.04),
 ('inland', 'clear', 1019.99),
 ('coastal', 'clear', 1019.85),
 ('coastal', 'partly cloudy', 1010.36),
 ('inland', 'partly cloudy', 1006.62),
 ('mountain', 'partly cloudy', 1006.48),
 ('coastal', 'overcast', 1006.3),
 ('mountain', 'cloudy', 1000.87),
 ('inland', 'cloudy', 998.36),
 ('mountain', 'overcast', 998.03),
 ('inland', 'overcast', 997.51),
 ('coastal', 'cloudy', 997.16)]

<a id="summary"></a>

## **3.4 Summary and Key Insights**

After running the SQL queries, we summarize the key findings and insights to guide the next phases of the project.

### **Summary**

This section presents a summary of the key findings from the weather dataset, analyzed using SQL queries in a SQLite environment. The goal was to understand how different atmospheric factors vary across seasons, weather types, and locations. Through structured queries, we uncovered several meaningful patterns related to temperature, precipitation, wind, UV exposure, and visibility.

#### 1. Weather Type Distribution and Visibility

The most frequent weather type is Snowy, followed by Cloudy, Rainy, and Sunny. Snowy conditions during Winter are associated with the lowest average visibility of 2.97 kilometers, suggesting reduced clarity and potentially hazardous conditions.

#### 2. Seasonal Temperature and Precipitation Trends

Spring has the highest average temperature at 25.35°C, slightly higher than Autumn and Summer. Winter remains the coldest season. In terms of precipitation, Winter experiences the highest levels, particularly during Snowy and Rainy conditions, each averaging over 76 mm.

#### 3. Atmospheric and Environmental Factors

Clear skies, especially in mountain and inland areas, are linked to higher average atmospheric pressure, indicating more stable weather conditions. In contrast, overcast and cloudy weather in coastal and inland locations are associated with lower pressures, which may indicate more unstable atmospheric activity.

#### 4. Wind and UV Patterns by Condition and Location

Rainy days record the highest average wind speed at 13.82 km/h, significantly more than other conditions. Coastal regions receive the highest average UV index at 4.37, indicating stronger solar radiation compared to inland and mountain regions.

### **Conclusion**

The SQL-driven exploration of the weather dataset reveals consistent patterns across various climate indicators. These insights provide a foundational understanding of how weather conditions shift across different seasons and geographies. They can inform local planning, environmental assessments, and further meteorological research.
